### PR TITLE
Add Matrix Chain Multiplication in Java to NeoAlgo

### LINKED ISSUE #3405: Matrix Chain Multiplication (Java, DP)
## 💥 Proposal

Matrix chain multiplication is an important optimization problem which is used to find the most efficient way to multiply a given sequence of matrices. Main problem is to decide the sequence of multiplying the given matrices in an optimal way and hence finding the optimal cost rather than actually performing the multiplication for the same. 
This can be implemented using 3 approaches, but I would like to implement it using Dynamic Programming, in which recursive algorithm will be used to break the problem into a set of related subproblems and eventually finding the minimum cost.

For implementing this, I'll be using *java*.

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes, I have read the guidelines.

### PR DESCRIPTION
**Pull request by leoshrey**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes, I have read the contribution guidelines.
### Description

This adds Matrix Chain Multiplication implementation in Java by following Dynamic Programming approach which leads to the optimal solution. 
### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Fixes #3405

### CODE REVIEW

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Few changes: 

- Add Multi-Line Comments 
- Add a beginner-friendly description at the top
- Use proper code Format 
- Remove extra lines of code and spaces
- Add sample I/O at the end of the Code
- Add Time and Space Complexity at the end of the Code

**Comment by leoshrey** [comment]
@ankitaggarwal23, changes are made successfully.

File: Java/dp/MatrixChainMultiplication.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+                        There are 4 matrices of dimensions 10x20, 20x30, 30x40 and 40x30.
+                        Let the input 4 matrices be A, B, C and D.  The minimum number of
+                        multiplications are obtained by putting parenthesis in following way
+                        ((AB)C)D --> 10*20*30 + 10*30*40 + 10*40*30
+*/
+/*  Time and space complexity for this approach:
+    Time Complexity: O(n^3)
+    Auxiliary Space: O(n^2)
+*/
```
Comment (bhav09) [inline Java/dp/MatrixChainMultiplication.java:84]: Add a new line at EOF

**Review by bhav09 — CHANGES_REQUESTED** [review]
Reviewed.

File: Java/dp/MatrixChainMultiplication.java (hunk 1 above)
Comment (leoshrey) [reply Java/dp/MatrixChainMultiplication.java:79]: sure

**Comment by leoshrey** [comment]
Actually, I have already left a new line at EOF, it cannot be seen on this commit page, as it only shows the lines with changes made. But it can be seen in the code.

File: Java/dp/MatrixChainMultiplication.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+	// Matrix Chain Multiplication Problem
+	public static void main(String[] args) {
+		int n;
+		Scanner sc = new Scanner(System. in );
+		System.out.print("Enter the number of matrices you want to multiply: ");
+		// Matrix `M[i]` has dimension `dimension[i-1] × dimension[i]` for `i = 1…n`
+		// Storing value 1 more than the entered number as dimensions
+		// As dimensions of 3 matrices are  `10 × 30` matrix, `30 × 5` matrix, `5 × 60` matrix
+		// in case of int[] dimension = { 10, 30, 5, 60 };
```
Comment (HarshCasper) [inline Java/dp/MatrixChainMultiplication.java]: Use multi-line comments

File: Java/dp/MatrixChainMultiplication.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		int minMult = Integer.MAX_VALUE;
+		// take the minimum over each possible position at which the
+		// sequence of matrices can be split
+		/*
+            (M[i+1]) × (M[i+2]………………M[j])
+            (M[i+1]M[i+2]) × (M[i+3…………M[j])
+            …
+            …
+            (M[i+1]M[i+2]…………M[j-1]) × (M[j])
+         */
```
Comment (HarshCasper) [inline Java/dp/MatrixChainMultiplication.java]: Use multi-line comments

File: Java/dp/MatrixChainMultiplication.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.Scanner;
+public class MatrixChainMultiplication {
+	// To find efficient way to multiply given matrices and hence finding optimal cost.
+	public static int matrixChain(int[] dimension, int i, int j) {
+		// case 1: when only one matrix is there
+		if (j <= i + 1) {
+			return 0;
+		}
+		// minMult : stores the minimum number of scalar multiplications (i.e., cost)
+		// needed to compute matrix `M[i+1] … M[j] = M[i…j]`
```
Comment (HarshCasper) [inline Java/dp/MatrixChainMultiplication.java]: Use multi-line comments

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by leoshrey** [comment]
@HarshCasper, Changes are made successfully

**Review by HarshCasper — APPROVED** [review]

**Review by ankitaggarwal23 — COMMENTED** [review]
Code plagiarized from [here](https://www.techiedelight.com/matrix-chain-multiplication/#:~:text=Matrix%20chain%20multiplication%20(or%20Matrix,of%20the%20matrix%20multiplications%20involved.)

**Pull request merged**